### PR TITLE
commands: support .sh commands

### PR DESCRIPTION
**Pull request by bfontaine**
I also refactored things a bit and added tests.

Without this change `update-bash.sh` appears in the `brew commands` output instead of `update-bash`.

Update: this also fixes an issue with the shell completion, right now `brew upd<tab>` gives:
```
update          update-bash.sh  update-report   update-test
```
Instead of:
```
update         update-bash    update-report  update-test
```

### CODE REVIEW

File: Library/Homebrew/cmd/commands.rb [hunk 1]
```diff
@@ -27,18 +27,30 @@ def commands
...
+      end
+      cmds
+    end.sort
+  end
+
+  private
+
+  def find_internal_commands(directory)
+    directory.children.reduce([]) do |cmds, f|
+      cmds << f.basename(".rb").to_s.sub(/\.sh$/, "") if f.file?
```
Comment (bfontaine) [inline Library/Homebrew/cmd/commands.rb]: Note that this won’t work if we have a command named e.g. `foo.sh` that lives in `cmd/foo.sh.rb` but that’s unlikely.

**Comment by bfontaine** [comment]
Updated with a similar update for the `command` command.

File: Library/Homebrew/cmd/commands.rb (hunk 1 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/commands.rb]: Why not:
```ruby
f.basename(".*") if f.file? && %w[.rb .sh].include?(f.extname)
```

Comment (mikemcquaid) [inline Library/Homebrew/cmd/commands.rb]: I think `foo.sh.rb` is a bad idea anyway :trollface:

**Comment by mikemcquaid** [comment]
Great work here :+1:. I believe you can also remove `--repository` now that this is ready and add it to a list of `internal commands` or something, as previously discussed. Feel free to leave that for another PR, though.

File: Library/Homebrew/cmd/commands.rb [hunk 2]
```diff
@@ -27,18 +27,30 @@ def commands
   end
 
   def internal_commands
-    (HOMEBREW_LIBRARY_PATH/"cmd").children.select(&:file?).map { |f| f.basename(".rb").to_s }
+    find_internal_commands HOMEBREW_LIBRARY_PATH/"cmd"
   end
 
   def internal_development_commands
-    (HOMEBREW_LIBRARY_PATH/"dev-cmd").children.select(&:file?).map { |f| f.basename(".rb").to_s }
+    find_internal_commands HOMEBREW_LIBRARY_PATH/"dev-cmd"
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/commands.rb]: Note that `bin/brew` currently doesn't support internal developer-only commands. Not sure if we should take that into account here or extend `bin/brew` to support them. Thoughts?

Comment (mikemcquaid) [inline Library/Homebrew/cmd/commands.rb]: I'd be easy either way.

Comment (UniqMartin) [inline Library/Homebrew/cmd/commands.rb]: I'm in favor of supporting them in `bin/brew` so there's no confusing difference between Ruby and Shell command support. Thus, this could stay as-is.

I wouldn't mind creating a PR for `bin/brew` to add `dev-cmd` support there, if I'm not the only one who thinks this is a good idea. (I should probably wait for #48197 to be merged first.)

File: Library/Homebrew/cmd/commands.rb (hunk 1 above)
Comment (bfontaine) [inline Library/Homebrew/cmd/commands.rb]: Alternatively:
```rb
f.basename.to_s.sub(/\.(?:rb|sh)$/ if f.file?
```

Comment (UniqMartin) [inline Library/Homebrew/cmd/commands.rb]: I think this has the possible downside of allowing arbitrary files in those directories to be listed as commands. Previously, it would have ignored everything with an unsupported file extension. That said, there are currently no non-command files in those directories so that's not an actual issue.

Comment (bfontaine) [inline Library/Homebrew/cmd/commands.rb]: > Previously, it would have ignored everything with an unsupported file extension.

I don’t think so. The previous code was:
```rb
(HOMEBREW_LIBRARY_PATH/"dev-cmd").children.select(&:file?).map { |f| f.basename(".rb").to_s }
```

I originally had more complicated code to avod this but I simplified it for this PR. If you edit command files and run `brew commands` at the same time you might end up with commands like `foo.rb~` and/or `.foo.rb.swp` if you use Vim. We could add a filter like `/^-*[a-z-]+$/` but this issue is unlikely to arise if you don’t touch the code so I’m not sure it’s necessary.

Comment (UniqMartin) [inline Library/Homebrew/cmd/commands.rb]: Sorry, you are right. (I was misguided by Xu's comment instead of looking at the original code.) I don't think the effort to further filter this is warranted, given this is only used with directories where we exercise full control (and shouldn't put unrelated stuff in there).

**Comment by bfontaine** [comment]
@mikemcquaid I’ll do that in a separate PR. Eventually I’d like to move the common code from `commands.rb` and `command.rb` in a core file, but I’m waiting for the new `update` mechanism to go wild since you said you’d prefer we don’t create new core files before that.

File: Library/Homebrew/cmd/commands.rb (hunk 1 above)
Comment (UniqMartin) [inline Library/Homebrew/cmd/commands.rb]: Just wanted to add that I'm happy with the suggested alternative, if you add the replacement argument and add a closing parenthesis. :wink: Maybe worth leaving out `?:` in the regexp, as it doesn't hurt to have the capture and makes the regexp less readable.

Comment (bfontaine) [inline Library/Homebrew/cmd/commands.rb]: Thanks, I used the suggested alternative. However I respectfully disagree on the `?:`; it doesn’t hurt *not* to have the capture :stuck_out_tongue: Unless you have a strong opinion on this I’ll keep it that way.

Comment (UniqMartin) [inline Library/Homebrew/cmd/commands.rb]: No strong opinion. I'm fine with the `?:` and it won't distract me. :smiley_cat:

**Comment by UniqMartin** [comment]
:+1: if the bots are happy.
